### PR TITLE
Coroutines Converstion: convert api kv.c++ to coroutines

### PR DESCRIPTION
**Pull request by jasnell**

### CODE REVIEW

**Comment by vlovich** [comment]
I'll defer to @mkevanz / @elithrar

File: src/workerd/api/kv.c++
Comment (mikea) [inline src/workerd/api/kv.c++:327]: do we actually need this ref counting now that this is a co-routine?

File: src/workerd/api/kv.c++
Comment (mikea) [inline src/workerd/api/kv.c++:332]: maybe drop writePromise and just put co_await at write() call? Will make it easier to change too.

**Review by mikea — APPROVED** [review]

**Comment by jasnell** [comment]
Going to move this to draft status so it doesn't get landed yet. We're going to hold off converting any of the kv stuff more until that team gives a green light.

File: src/workerd/api/kv.c++
Comment (kentonv) [inline src/workerd/api/kv.c++:315]: I feel that we should almost never use `auto` parameters to a function like this, because it takes away all ability for the IDE the understand what they are. Auto-complete, jump-to-definition, and error detection are all broken. As a human, to figure out the types of these parameters I have to scroll all the way down to the call site.

But, of course, writing out the full type is tedious and verbose.

I also feel like these functions break up the flow of the code in an unnatural way. Factoring out functions improves code readability when the function provides a natural abstraction. But in this case there's no abstraction, it's just "the rest of the stuff that has to be done after the I/O call returns". I think it just makes things harder to read by making me scroll up and down to see how the control flows.

So I think the ideal solution is to stick with lambdas and lambda captures instead of factoring out into a function like this. The you don't have to write out all the types, but the compiler still knows what they are. And then the code stays together naturally in one place.

But then, of course, we have the lambda coroutine problem.

This is where I think `kj::coCapture()` makes sense to use. Can we just use that?

Perhaps we could go so far as to make `context.awaitIo()` use `coCapture()` automatically?

Of course, when we have coroutines for `jsg::Promise`, then we can just make it possible to `co_await` a KJ promise directly without using `awaitIo`, and the lambda can fully go away.

**Pull request closed without merging**
